### DISCUSSION TITLE
cexternal: the proposal

### DISCUSSION THREAD

**Discussion by samiam95124** [samiam95124/Pascal-P6#111] [General]
### A proposed local extention: cexternal

This is an expansion on the idea introduced in the last post: "cross calling C".

First, what is a "local extension"? That is a feature exclusive to Pascal-P6, and not part of the Pascaline specification. It's that way because it implements a feature that is specific to the way Pascal-P6 works.

cexternal is proposed as an addition to the existing reserved-word "external". External is defined in Pascaline as a reserved word, but its exact function is not specified. It is implemented in Pascal-P6 and has the following function:

A procedure or function marked as external is used to satisfy external references, but has no body, and further, suppresses errors as to:

1. Use, or non-use of parameters.
2. Lack of assignment to any function result.

The proposed cexternal has the same properties as external, but adds the following:

1. A container array of base type char and is packed that is passed as a value parameter will be converted to a zero terminated string by copying the contents to a buffer of its size plus one, then it has a zero word appended to it. The word size of the terminator matches the UTF status, that is, UTF-8 is 8 bits, UTF-16 is 16 bits, UTF-32 is 32 bits. The zero terminated string is then passed. Note that var passed containers do not have this conversion, nor do fixed (ISO 7185) arrays of character, nor arrays that do not qualify as strings (not packed, not based on char, not having a beginning index of 1)[1]. Note that the predefined Pascaline type "string", when passed as value, qualifies.
2. A pointer type of any base type will accept, as a parameter, any of the following:

A. A matching pointer type of the same base (same as ISO 7185 rules).
B. Any variable of the same base type as the pointer parameter. In this case, the result is of taking the address of that variable and passing that[2].

### Purpose of cexternal

cexternal will dramatically reduce the work of interfacing with the C language. The rule to pass as zero terminated allows conversion to a zero terminated string and is relatively easy to bypass (pass as var). The pointer rule covers the fact that C is ambiguous about what the meaning of a pointer is with respect to a parameter. In C a pointer parameter can function as any of:

1. A pointer.
2. An array.
3. A variable passed parameter (can be modified by the function called).

### TL;DR

The cexternal system basically leaves it up to the programmer as to the proper use of a C function call. It is not safe or protected, but then neither is code written in C. Its confined to C and calls to C, and can't be used in normal Pascaline code (at least not without some gymnastics). Further, programmers wanting to keep the type safe status of Pascaline even in C programs can use the regular external call formatting. Then either the called C program is compliant with Pascaline calling format, or a wrapper that is is used as an intermediary to the code.

The alternative to this method using ch2ph is to pepper the file with anotations designed to produce a "safe" interpretation of the C function call. This method avoids that, at the cost of more work for the programmer to avoid misuse of the call.

[1] I believe view parameters will be equivalent to value in this respect, but I am still considering it.
[2] I understand that pointer to pointer is a mildly ambiguous rule.

**Comment by samiam95124** [comment]
I'm going to critique my own proposal here.

The method described is clever, and does save a lot of work in writing annotations.

However, it basically imports fairly dirty interfaces into Pascaline. Even though it is limited to C calls (and it *could* be used for Pascaline calls if you really tried), the other methods outside of cexternal at least have their type safety issues in the C code. The cexternal proposal would import them to the Pascaline side.

Further, the work that is being saved here is that of carefully going over imported C APIs and writing annotations, wrappers or a combination of both. The result is a type safe API. I still thing that is the right way to go, and if the C API is worth importing, its worth the work.

All of this, thankfully, is not going to be an issue until ch2ph gets ported to Pascal-P6. In the meantime, I don't want to create false expectations, so I am closing this discussion.
